File: internal/fsvm/fsvm.go
package fsvm

// FSVM: Fibonacci Stream Virtual Machine
//
// Deterministic, measurement-first streaming core.
// Tracks:
// - r: dilation exponent (retrospective upsample-by-2 count)
// - w: 6-bit hexagram window (recent logical bits)
// - zeroRun, lastBit
// - sketch: Zobrist state sketch (XOR-folded, for cheap divergence detection)
//
// Dilation protocol:
// - if observed adjacency 11 occurs, emit DILATE and increment r.
//
// Segmentation is NOT decided here; it is an interpretation layer (NFA) driven by markers.

type EventKind uint8

const (
	EventDilate EventKind = iota + 1
	EventMarker
)

type Event struct {
	Kind    EventKind
	Payload uint64
}

// DefaultSeeds is the standard Zobrist seed table.
// Index 0 for bit=0, index 1 for bit=1.
var DefaultSeeds = [2]uint64{
	0x517cc1b727220a95, // bit=0
	0x9e3779b97f4a7c15, // bit=1
}

type State struct {
	Seeds      [2]uint64 // per-instance Zobrist seed table
	MixA       uint64    // hash-family multiplier (v2)
	MixB       uint64    // hash-family addend (v2)
	MixR       uint8     // hash-family rotation (v2)
	Sketch     uint64    // Zobrist state sketch
	SketchDelta uint8    // rolling bits-changed in sketch (v2)
	ZeroRun    uint64
	Dilations  uint64
	Markers    uint64
	R          uint32 // dilation exponent
	W          uint8  // 6-bit window
	LastBit    uint8
	Width      uint8  // adjacency width (used by calibration/step-width)
}

// New returns a State with default seeds.
func New() State {
	return State{Seeds: DefaultSeeds}
}

// NewWithSeeds returns a State with caller-provided seeds.
func NewWithSeeds(seeds [2]uint64) State {
	return State{Seeds: seeds}
}

func isPow2(x uint64) bool { return x > 0 && (x&(x-1)) == 0 }

// Step consumes one observed bit and returns zero or more events.
//
// Policy (measurement-first, unDoSable):
// - Adjacent 1s => DILATE (r++).
// - Zero-run markers emitted at sparse powers-of-two crossings >= 8: 8,16,32,...
func Step(s State, b uint8) (State, []Event) {
	b &= 1
	var evs []Event

	// update zero run + marker
	if b == 0 {
		s.ZeroRun++
		if s.ZeroRun >= 8 && isPow2(s.ZeroRun) {
			s.Markers++
			evs = append(evs, Event{Kind: EventMarker, Payload: s.ZeroRun})
		}
	} else {
		s.ZeroRun = 0
	}

	// detect adjacency
	if s.LastBit == 1 && b == 1 {
		s.R++
		s.Dilations++
		evs = append(evs, Event{Kind: EventDilate, Payload: uint64(s.R)})
	}

	s.LastBit = b
	s.W = ((s.W << 1) | b) & 0x3F

	// Zobrist fold: one XOR using per-instance seed, fused into hot path.
	// Fold both bit value and window state to avoid sketch collapse on
	// streams with equal 0/1 counts (XOR self-inverse property).
	s.Sketch ^= s.Seeds[b] + uint64(s.W)

	return s, evs
}

File: scripts/python_ref/fsvm.py
#!/usr/bin/env python3
"""fsvm.py — FSVM reference core (Python)

Extends the earlier `fsvm_ref.py` into a reusable module.

Core rules:
- Observe boolean stream.
- If adjacent 1s occur, emit DILATE and increment r.
- Track a 6-bit sliding window (hexagram) and zero-run length.

Marker emission (optional, unDoSable default):
- emit MARKER when zero-run crosses powers of two: 8,16,32,...

This keeps candidate segmentation points sparse.
"""

from __future__ import annotations

from dataclasses import dataclass


def _is_pow2(x: int) -> bool:
    return x > 0 and (x & (x - 1)) == 0


@dataclass
class Event:
    kind: str
    payload: int = 0
    i: int = 0


@dataclass
class State:
    r: int = 0
    w: int = 0      # 6-bit
    last: int = 0
    zero_run: int = 0
    dilations: int = 0
    markers: int = 0


def step(st: State, b: int, i: int) -> list[Event]:
    b &= 1
    evs: list[Event] = []

    # zero run + markers
    if b == 0:
        st.zero_run += 1
        # sparse markers at 8,16,32,... (not at 1,2,4)
        if st.zero_run >= 8 and _is_pow2(st.zero_run):
            st.markers += 1
            evs.append(Event(kind="MARKER", payload=st.zero_run, i=i))
    else:
        st.zero_run = 0

    # adjacency detection
    if st.last == 1 and b == 1:
        st.r += 1
        st.dilations += 1
        evs.append(Event(kind="DILATE", payload=st.r, i=i))

    st.last = b
    st.w = ((st.w << 1) | b) & 0x3F
    return evs
